x/stakeibc/keeper/validator_selection.go

```diff
@@ -4,60 +4,255 @@ import (
+	// Get the host zone and confirm the delegation account is initialized
+	hostZone, found := k.GetHostZone(ctx, chainId)
+	if !found {
+		return errorsmod.Wrap(types.ErrHostZoneNotFound, fmt.Sprintf("Host zone %s not found", chainId))
+	}
+	if hostZone.DelegationIcaAddress == "" {
+		return errorsmod.Wrapf(types.ErrICAAccountNotFound, "no delegation account found for %s", chainId)
+	}
+
+	// Get the difference between the actual and expected validator delegations
+	valDeltaList, err := k.GetValidatorDelegationDifferences(ctx, hostZone)
+	if err != nil {
+		return errorsmod.Wrapf(err, "unable to get validator deltas for host zone %s", chainId)
+	}
+
+	msgs, rebalacings := k.GetRebalanceICAMessages(hostZone, valDeltaList, numRebalance)
+
+	// marshall the callback
+	rebalanceCallback := types.RebalanceCallback{
+		HostZoneId:   hostZone.ChainId,
+		Rebalancings: rebalacings,
+	}
+	rebalanceCallbackBz, err := k.MarshalRebalanceCallbackArgs(ctx, rebalanceCallback)
+	if err != nil {
+		return errorsmod.Wrapf(err, "unable to marshal rebalance callback args")
+	}
+
+	// Submit the rebalance ICA
+	_, err = k.SubmitTxsStrideEpoch(
+		ctx,
+		hostZone.ConnectionId,
+		msgs,
+		types.ICAAccountType_DELEGATION,
+		ICACallbackID_Rebalance,
+		rebalanceCallbackBz,
+	)
+	if err != nil {
+		return errorsmod.Wrapf(err, "Failed to SubmitTxs for %s, messages: %+v", hostZone.ChainId, msgs)
+	}
+
+	return nil
+}
+
+// Given a list of target delegation changes, builds the individual re-delegation messages by redelegating
+// from overweight validators to underweight validators
+// Returns the list of messages and the callback data for the ICA
+func (k Keeper) GetRebalanceICAMessages(
+	hostZone types.HostZone,
+	validatorDeltas []RebalanceValidatorDelegationChange,
+	numRebalance uint64,
+) (msgs []sdk.Msg, rebalancings []*types.Rebalancing) {
+	// Sort the list of delegation changes by the size of the change
+	lessFunc := func(i, j int) bool {
+		if !validatorDeltas[i].Delta.Equal(validatorDeltas[j].Delta) {
+			return validatorDeltas[i].Delta.LT(validatorDeltas[j].Delta)
+		}
+		// use name as a tie breaker if deltas are equal
+		return validatorDeltas[i].ValidatorAddress < validatorDeltas[j].ValidatorAddress
+	}
+	sort.SliceStable(validatorDeltas, lessFunc)
+
+	// QUESTION: Does anyone have a preference on the naming convention here? A couple options:
+	//  * Overweight/Underweight (current)
+	//  * Surplus/Deficit
+	//  * Source/Destination
+	//  * Giver/Receiver
+	// etc.
+
+	// Pair overweight and underweight validators, with a redelegation from the overweight
+	// validator to the underweight one
+	// The list is sorted with the overweight validators (who should lose stake) at index 0
+	//   and the underweight validators (who should gain stake) at index N-1
+	// The overweight validator's have a negative delta and the underweight validators have a positive delta
```

Comment:
imo it's a little bit unintuitive that Surplus Validators have a negative delta. Should we invert the deltas, so define Surplus Validator as having positive deltas, and Deficit Validators as having negative deltas?

Reply:
yeah I thought the same thing! Will do

Reply:
One other option is to just have them all be positive and add a field to the Delta struct that says whether it was a surplus or deficit.

Not sure if this would clean it up or just make it more congested, but would be nice to not have to think about the sign at all!

Reply:
nvm, I think swapping the sign cleaned it up plenty

Reply:
Yeah I'm in favor of swapping signs for now! Not opposed to adding a third field too if you think it's still confusing though, esp given the density of this code

Reply:
reading over the code again, I don't think the additional field would make it any clearer, but we can always come back to it! Marking this resovled